[SPARK-160] Configurable Mesos framework webui URL.

s-urbaniak opened the pull request:
Previously the Mesos framework webui URL was being derived only from the
Spark UI address leaving no possibility to configure it. This commit
makes it configurable. If unset it falls back to the previous behavior.

mgummelt commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala:
```diff
@@ -73,7 +73,7 @@ private[mesos] class MesosClusterDispatcher(
   def start(): Unit = {
     webUi.bind()
-    scheduler.frameworkUrl = webUi.activeWebUiUrl
+    scheduler.frameworkUrl = conf.getOption("spark.mesos.webui.url").getOrElse(webUi.activeWebUiUrl)
```
It looks like the standard in this code base is to use `get(config, default)`

mgummelt commented:
Can you add tests?  I'd like to have unit tests for each PR we send.

mgummelt commented on docs/running-on-mesos.md:
```diff
@@ -387,6 +387,14 @@ See the [configuration page](configuration.html) for information on Spark config
+  <td><code>spark.mesos.webui.url</code></td>
+  <td><code>(none)</code></td>
+  <td>
+    Set the Spark Mesos framework's webui_url pointing to a web-based interface for interacting with the framework.
```
s/pointing/to point

"to point a web-based interface for interacting with the framework" is a bit redundant.  we can probably remove it.

Not a blocker

s-urbaniak commented:
@mgummelt addressed review comments. I am not sure of what we actually want to unit test here? Is it that the backends and the dispatcher pick up the "correct" webui.url settings? We'd unit test the `SparkConf` `get` methods rather than the backends/dispatchers which wouldn't make sense in this context.

I'd rather see an e2e-style test here but am a little lost where to put this kind of test.

s-urbaniak commented:
PS: Did a code coverage over the existing backend/dispatcher test suites and exactly those start methods are not being covered right now. I could factor the initialization of the `web.uri` as class members and validate the state in the unit tests, but this doesn't sound right either (asserting internal object state inside an unit test).

mgummelt commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -136,7 +136,8 @@ private[spark] class CoarseMesosSchedulerBackend(
       sc.appName,
       sc.conf,
-      sc.ui.map(_.appUIAddress))
+      sc.conf.getOption("spark.mesos.webui.url").orElse(sc.ui.map(_.appUIAddress))
```
These should all be .get(val, default), right?

s-urbaniak replied:
well, `sc.ui.map(...)` is an Option construct itself, thus this construct. `orElse` accepts an Option like `flatMap`, whereas `.get(val, default)` expects an explicit value.

mgummelt replied:
both result in returning an Option[String], right?

Regardless, not important.

mgummelt replied:
Oh, nevermind.  I see.

mgummelt commented:
A reasonable unit test would be that e.g. scheduler.frameworkuri is set appropriately after calling start()

mgummelt commented:
Don't let the unit tests block you though.  This LGTM to submit upstream

tnachen commented on docs/running-on-mesos.md:
```diff
@@ -387,6 +387,14 @@ See the [configuration page](configuration.html) for information on Spark config
   </td>
 </tr>
+<tr>
+  <td><code>spark.mesos.webui.url</code></td>
```
I think I understand the intent, but the configuration setting here seems to be too general.
In the cases you changed the web ui, in the coarse/fine grain scheduler you're replacing the Spark App UI which is the default per Driver UI. In the Dispatcher case you're replacing the Dispatcher framework UI.
I think we want to be more specific here, one setting for the dispatcher uri another for the driver UI, which actually can be a generic setting not just for Mesos.

mgummelt replied:
I would agree if running the dispatcher and running the driver weren't mutually exclusive.  But since you can only run one at a time, I think it makes sense to just have one config var.  It's like the "docker.image" config var.  It can refer to the driver docker image or the executor docker image.

s-urbaniak commented:
/cc @mgummelt

mgummelt commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -136,7 +136,8 @@ private[spark] class CoarseMesosSchedulerBackend(
       sc.appName,
       sc.conf,
-      sc.ui.map(_.appUIAddress))
+      sc.conf.getOption("spark.mesos.scheduler.webui.url").orElse(sc.ui.map(_.appUIAddress))
```
Can you call this "driver" instead of "scheduler"?  The dispatcher is also a Mesos "scheduler", so this is ambiguous.

mgummelt commented:
One naming complaint, then LGTM, though I would still like to see tests

s-urbaniak commented:
@mgummelt I am sort of lost:

- `CoarseMesosSchedulerBackendSuite`: doesn't start at all for me with a `Can't assign requested address: Service 'sparkDriver' failed after 16 retries! 
java.net.BindException` so I cannot really test `CoarseMesosSchedulerBackend`

- `MesosClusterDispatcher` is very much up in the architectual layers starting a jetty and binding ui ports and has no test suite at all so far.

- `MesosSchedulerBackendSuite` this one I could implement a test for, see latest commit :-)

mgummelt commented on core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendSuite.scala:
```diff
@@ -42,6 +42,41 @@ import org.apache.spark.scheduler.cluster.ExecutorInfo
+    conf.set("spark.mesos.driver.webui.url", "http://webui")
+    conf.set("spark.app.name", "name1")
+
+    val sc = mock[SparkContext]
```
There's already a `sc` in scope via `LocalSparkContext`.  Use that one.  See the other test suites for examples.

mgummelt commented:
Not sure about `CoarseMesosSchedulerBackendSuite`.  It works for me.

For `MesosClusterDispatcher`, it looks like you could mock `webUi` and `server`, then `start()` should be testable.  Again, not a big deal if you can't get this working.

`MesosSchedulerBackendSuite` LGTM

skonto commented:
What is the behavior if the uri is not correct, eg not valid hostname... does the ui fail or is this virtual?

s-urbaniak commented:
The uri being configured here is passed through as-is to the underyling `webui-url` field of `FrameworkInfo`, see https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L253. If the user configures bogus entries they will be simply passed through to `mesos.proto` and thus will appear as configured when the `state[.json]` endpoint is being invoked on Mesos.

These settings have no influence on the UI itself if that is what you mean by "virtual".

skonto commented:
Ok yes i see they are just to register frameworks webui uri. cool.

skonto commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -35,6 +36,11 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
   propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
 
   private def parse(args: List[String]): Unit = args match {
+    case ("--webui-url" | "-u") :: value :: tail =>
```
Need to update usage function at the end of the file with --webui-url

s-urbaniak replied:
Good catch, thanks!

skonto replied:
case ("--port" | "-p") :: IntParam(value) :: tail =>
      port = value
      parse(tail)

    case ("--webui-port" | "-p") :: IntParam(value) :: tail =>
      webUiPort = value
      parse(tail)
Same short  options for both: -p. This is not correct these are different ports. I propose not to have a short option like the --name long option which does not have a short one.

skonto replied:
Also -u is not so intuitive (probably it means ui). I would propose just to omit it.

s-urbaniak commented:
merged upstream
